torch/utils/data/dataloader.py

```diff
@@ -426,80 +542,77 @@ class _DataLoaderIter(object):
-                          self.worker_result_queue, self.done_event,
-                          self.collate_fn, base_seed + i,
-                          self.worker_init_fn, i))
-                w.daemon = True
-                # NB: Process.start() actually take some time as it needs to
-                #     start a process and pass the arguments over via a pipe.
-                #     Therefore, we only add a worker to self.workers list after
-                #     it started, so that we do not call .join() if program dies
-                #     before it starts, and __del__ tries to join but will get:
-                #     AssertionError: can only join a started process.
-                w.start()
-                self.index_queues.append(index_queue)
-                self.workers.append(w)
-
-            if self.pin_memory:
-                self.data_queue = queue.Queue()
-                pin_memory_thread = threading.Thread(
-                    target=_utils.pin_memory._pin_memory_loop,
-                    args=(self.worker_result_queue, self.data_queue,
-                          torch.cuda.current_device(), self.done_event))
-                pin_memory_thread.daemon = True
-                pin_memory_thread.start()
-                # Similar to workers (see comment above), we only register
-                # pin_memory_thread once it is started.
-                self.pin_memory_thread = pin_memory_thread
-            else:
-                self.data_queue = self.worker_result_queue
-
-            _utils.signal_handling._set_worker_pids(id(self), tuple(w.pid for w in self.workers))
-            _utils.signal_handling._set_SIGCHLD_handler()
-            self.worker_pids_set = True
+        super(_MultiProcessingDataLoaderIter, self).__init__(loader)
+
+        assert self.num_workers > 0
+
+        self.worker_init_fn = loader.worker_init_fn
+        self.worker_queue_idx_cycle = itertools.cycle(range(self.num_workers))
+        self.worker_result_queue = multiprocessing.Queue()
+        self.worker_pids_set = False
+        self.shutdown = False
+        self.send_idx = 0  # idx of the next task to be sent to workers
+        self.rcvd_idx = 0  # idx of the next task to be returned in __next__
+        # information about data not yet yielded, i.e., tasks w/ indices in range [rcvd_idx, send_idx).
+        # map: task idx => - (worker_id,)        if data isn't fetched (outstanding)
+        #                  \ (worker_id, data)   if data is already fetched (out-of-order)
+        self.task_info = {}
+        self.tasks_outstanding = 0  # always equal to count(v for v in task_info.values() if len(v) == 1)
+        self.done_event = multiprocessing.Event()
+
+        self.index_queues = []
+        self.workers = []
+        # A list of booleans representing whether each worker has exhausted
+        # its iterable dataset object. Always contain all Trues if not using
+        # an iterable dataset (i.e., if mode != Iterable).
```

Comment:
This comment is highly misleading, because this list actually holds a negation of this statement for each worker.